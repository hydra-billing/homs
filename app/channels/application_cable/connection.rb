module ApplicationCable
  class Connection < ActionCable::Connection::Base
    identified_by :current_user

    def connect
      self.current_user = find_verified_user
    end

    private

    def find_verified_user
      if env['warden'].user.nil?
        reject_unauthorized_connection
      else
        env['warden'].user
      end
    end
  end
end
