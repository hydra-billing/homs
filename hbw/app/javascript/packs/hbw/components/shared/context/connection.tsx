import React, { createContext, ComponentType, FC } from 'react';
import Connection from '../../../connection';

export type ContextType = {
  options: {
    host: string,
    path: string
  },
  request: Function,
  serverURL: string
}

const ConnectionContext = createContext<ContextType | null>(null);

type InitParams = {
  connection: Connection
}

export const withConnectionContext = (
  { connection }: InitParams
) => (WrappedComponent: ComponentType) => {
  const ConnectionProvider: FC = props => (
      <ConnectionContext.Provider value={connection} >
        <WrappedComponent {...props} />
      </ConnectionContext.Provider >
  );

  return ConnectionProvider;
};

export default ConnectionContext;
