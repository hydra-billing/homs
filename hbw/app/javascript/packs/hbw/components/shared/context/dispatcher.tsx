import React, { createContext, ComponentType, FC } from 'react';
import Dispatcher from '../../../dispatcher';

export type ContextType = Dispatcher;

const DispatcherContext = createContext<Dispatcher | null>(null);

type InitParams = { dispatcher: Dispatcher };

export const withDispatcherContext = (
  { dispatcher }: InitParams
) => (WrappedComponent: ComponentType) => {
  const DispatcherProvider: FC = props => (
    <DispatcherContext.Provider value={dispatcher} >
      <WrappedComponent {...props} />
    </DispatcherContext.Provider >
  );

  return DispatcherProvider;
};

export default DispatcherContext;
