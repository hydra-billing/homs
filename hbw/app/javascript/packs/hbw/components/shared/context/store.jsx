/* eslint-disable no-await-in-loop */
/* eslint-disable no-restricted-syntax */

import React, { Component, createContext } from 'react';
import ActionCable from 'actioncable';
import orderBy from 'lodash-es/orderBy';
import partition from 'lodash-es/partition';
import Messenger from 'messenger';
import { parseISO } from 'date-fns';

export const StoreContext = createContext({});
export const { Consumer: StoreConsumer } = StoreContext;

const withStoreContext = (WrappedComponent) => {
  class StoreProvider extends Component {
    state = {
      tasks:      [],
      events:     [],
      fetching:   true,
      ready:      false,
      socket:     null,
      error:      null,
      activeTask: null,
    };

    componentDidMount () {
      this.initSocket();
      this.initStore();
    }

    getTaskById = async (taskId, cacheKey = null) => {
      const { env } = this.props;

      const incomingTask = await env.connection.request({
        url:    `${env.connection.serverURL}/tasks/${taskId}`,
        method: 'GET',
        data:   {
          cache_key:    cacheKey,
          entity_class: env.entity_class,
        },
      }).then(response => response.json());

      const tasks = this.state.tasks.filter(task => task.id !== incomingTask.id);

      this.setState({
        tasks: this.orderTasks([...tasks, incomingTask])
      });

      return incomingTask;
    };

    onCreate = async (taskId, cacheKey) => {
      await this.getTaskById(taskId, cacheKey);
    };

    onAssignment = async (taskId, cacheKey, assignedToMe) => {
      const { translator: t } = this.props.env;
      const { name, assignee } = await this.getTaskById(taskId, cacheKey);

      if (assignee !== null) {
        if (assignedToMe) {
          Messenger.notice(t('notifications.new_assigned_task', {
            task_name: name
          }));
        } else {
          this.removeTaskFromList(taskId);
        }
      }
    };

    mergeFormsToTasks = (prevTasks, fetchedForms) => {
      const tasks = prevTasks;
      const forms = fetchedForms;

      Object.entries(tasks).forEach(([, value]) => {
        const fetchedForm = forms.find(form => form.task_id === value.id);

        if (fetchedForm !== undefined) {
          value.form = fetchedForm;
        }
      });

      return tasks;
    };

    getFormsForTasks = async (tasks) => {
      const emptyFormTasks = this.getEmptyFormEntityTasks(tasks);
      const { connection } = this.props.env;

      if (emptyFormTasks.length > 0) {
        const { forms } = await connection.request({
          url:    `${connection.serverURL}/tasks/forms`,
          method: 'GET',
          data:   {
            entity_tasks: JSON.stringify(emptyFormTasks)
          }
        }).then(data => data.json());

        this.addFormsToTasks(forms);
      }
    };

    getEmptyFormEntityTasks = (entityTasks) => {
      const entityClass = this.props.env.entity_class;

      const emptyFormEntityTasks = entityTasks.filter(
        task => (task.form === undefined)
      );

      return Object.entries(emptyFormEntityTasks).map(([, task]) => ({ task_id: task.id, entity_class: entityClass }));
    };

    addFormsToTasks = (fetchedForms) => {
      this.setState(prevState => ({
        tasks: this.mergeFormsToTasks(prevState.tasks, fetchedForms)
      }));
    };

    onComplete = async (taskId) => {
      this.removeTaskFromList(taskId);
    };

    onReceive = async ({
      task_id: taskId, cache_key: cacheKey, event_name: eventName, assigned_to_me: assignedToMe
    }) => {
      this.setState({ fetching: true });

      if (eventName === 'create') {
        await this.onCreate(taskId, cacheKey);
      }

      if (eventName === 'assignment') {
        await this.onAssignment(taskId, cacheKey, assignedToMe);
      }

      if (['complete', 'delete'].includes(eventName)) {
        await this.onComplete(taskId);
      }

      this.setState({ fetching: false });
    };

    defer = (callback) => {
      this.setState(prevState => ({ events: [...prevState.events, callback] }));
    };

    executeDeferred = async () => {
      const { events } = this.state;

      for (const e of events) {
        await e();
      }

      this.setState({ events: [] });
    };

    initSocket = () => {
      const { userIdentifier, widgetURL } = this.props.env;
      const { host, protocol } = new URL(widgetURL);

      const socketUrl = protocol === 'https:'
        ? `wss://${host}/widget/cable`
        : `ws://${host}/widget/cable`;

      const ws = ActionCable.createConsumer(socketUrl);

      ws.subscriptions.create({ channel: 'TaskChannel', user_identifier: userIdentifier }, {
        received: (data) => {
          const callback = async () => {
            await this.onReceive(JSON.parse(data));
          };

          if (this.state.ready) {
            callback();
          } else {
            this.defer(callback);
          }
        }
      });

      this.setState({ socket: ws });
    };

    initStore = async () => {
      const { env } = this.props;
      try {
        const result = await env.connection.request({
          url:    `${env.connection.serverURL}/tasks`,
          method: 'GET',
          data:   {
            entity_class: env.entity_class,
          },
        });

        const { tasks } = await result.json();

        this.setState({ tasks: this.orderTasks(tasks) });
      } catch (error) {
        this.setState({ error });
      } finally {
        this.setState({
          fetching: false,
          ready:    true
        }, this.executeDeferred);
      }
    };

    removeTaskFromList = async (taskId) => {
      const tasks = this.state.tasks.filter(task => task.id !== taskId);
      this.setState({ tasks });

      this.closeTaskOverview(taskId);
    };

    orderTasks = tasks => orderBy(
      tasks,
      [
        ({ due }) => (due ? parseISO(due) : null),
        'priority',
        'process_name',
        ({ created }) => (created ? parseISO(created) : null)
      ],
      ['asc', 'desc', 'asc', 'asc']
    );

    openTaskOverview = (task) => {
      const { activeTask } = this.state;

      if (!activeTask || task.id !== activeTask.id) {
        this.setState({ activeTask: task });
      }
    };

    closeTaskOverview = (taskId) => {
      const { activeTask } = this.state;

      if ((activeTask && taskId === activeTask.id) || !taskId) {
        this.setState({ activeTask: null });
      }
    };

    updateContext = (...state) => {
      this.setState(...state);
    };

    render () {
      const [unassignedTasks, myTasks] = partition(this.state.tasks, { assignee: null });
      const count = {
        my:         myTasks.length,
        unassigned: unassignedTasks.length,
      };

      const contextValue = {
        ...this.state,
        myTasks,
        unassignedTasks,
        count,
        update:           this.updateContext,
        openTask:         this.openTaskOverview,
        closeTask:        this.closeTaskOverview,
        getFormsForTasks: this.getFormsForTasks
      };

      return (
        <StoreContext.Provider value={contextValue}>
          <WrappedComponent {...this.props} />
        </StoreContext.Provider>
      );
    }
  }

  return StoreProvider;
};

export default withStoreContext;
