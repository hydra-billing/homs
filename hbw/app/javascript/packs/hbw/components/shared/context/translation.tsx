import React, {
  createContext, ComponentType, FC, useState, useEffect, useContext
} from 'react';
import TranslatorFactory from '../../../translator';
import { localizer } from '../../../init/date_localizer';
import ConnectionContext, { ContextType as ConnectionContextType } from './connection';

export type ContextType = {
  locale: {
    code: 'en' | 'ru' | 'es' | 'fr',
    dateTimeFormat: string
  }
  translate: (key: string, vars?: object, fallback?: string) => string,
  translateBP: (key: string, vars?: object, fallback?: string) => string,
  Localizer: {
    localizeDatetime: (str: string) => string | null,
    localizeDayMonth: (str: string) => string | null,
    localizeDayMonthYear:(str: string) => string | null
  }
}

const TranslationContext = createContext<ContextType | null>(null);

type InitParams = {
  locale: {
    code: 'en' | 'ru' | 'es' | 'fr',
    dateTimeFormat: string
  }
}

export const withTranslationContext = (
  { locale }: InitParams
) => (WrappedComponent: ComponentType) => {
  const TranslationProvider: FC = (props) => {
    const connection = useContext(ConnectionContext) as ConnectionContextType;

    const [BPTranslator, setBPTranslator] = useState({ t: () => '' });

    useEffect(() => {
      const fetchBPTranslations = async () => {
        const response = await connection.request({
          url:    `${connection.serverURL}/translations`,
          method: 'GET'
        });

        const BPTranslations = await response.json();

        setBPTranslator({ t: TranslatorFactory.getTranslatorForLocale(locale.code, BPTranslations) });
      };

      fetchBPTranslations();
    }, []);

    const translate = TranslatorFactory.getTranslatorForLocale(locale.code);
    const Localizer = localizer(locale);

    return (
      <TranslationContext.Provider value={{
        locale, translate, translateBP: BPTranslator.t, Localizer
      }}>
        <WrappedComponent {...props} />
      </TranslationContext.Provider>
    );
  };

  return TranslationProvider;
};

export default TranslationContext;
