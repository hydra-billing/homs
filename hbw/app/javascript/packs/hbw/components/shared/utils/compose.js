export default (...args) => value => args.reduceRight((acc, func) => func(acc), value);
