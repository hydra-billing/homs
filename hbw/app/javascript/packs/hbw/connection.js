import qs from 'qs';

export default class Connection {
  constructor (options) {
    this.options = options;

    if (!this.options.path) {
      this.options.path = '';
    }

    if (!this.options.host) {
      this.options.host = `${document.location.protocol}//${document.location.host}`;
    }

    this.serverURL = this.options.host + this.options.path;
  }

  request = (opts) => {
    if (opts.method.toLowerCase() === 'get') {
      const url = new URL(opts.url, window.location);

      const searchParams = [...url.searchParams]
        .reduce((obj, [key, value]) => ({ ...obj, [key]: value }), {});

      url.search = qs.stringify({
        ...searchParams,
        ...(opts.data || {}),
        payload: this.options.payload
      });

      return fetch(url.href);
    } else {
      return fetch(opts.url, {
        ...opts,
        body: JSON.stringify({
          ...opts.data,
          payload: this.options.payload
        })
      });
    }
  };
}
