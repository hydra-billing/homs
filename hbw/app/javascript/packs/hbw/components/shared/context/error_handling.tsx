import Messenger from 'messenger';
import React, {
  createContext, ComponentType, FC, useState, useCallback, useContext
} from 'react';
import HBWError from '../../error';
import TranslationContext, { ContextType as TranslationContextType } from './translation';
import ConnectionContext, { ContextType as ConnectionContextType } from './connection';
import { CamundaError, WidgetErrorType } from '../utils/errors';

export type ContextType = {
  addError: Function,
  formError: React.JSX.Element,
  safeRequest: Function
}

const ErrorHandlingContext = createContext<ContextType | null>(null);

export const withErrorHandlingContext = () => (WrappedComponent: ComponentType) => {
  const ErrorHandlingProvider: FC = (props) => {
    const { translate: t } = useContext(TranslationContext) as TranslationContextType;
    const { request } = useContext(ConnectionContext) as ConnectionContextType;

    const [formError, setFormError] = useState<React.JSX.Element>(() => <HBWError/>);

    const notify = useCallback((error: WidgetErrorType) => {
      // @ts-ignore
      Messenger[error.level](t(error.notificationText, {}, error.notificationText));
    }, [t]);

    const setupFormError = useCallback((error: WidgetErrorType) => {
      setFormError(
        <HBWError
          errorHeader={t(error.header, {}, error.header)}
          errorBody={error.body}
        />,
      );
    }, [t]);

    const addError = useCallback((error: WidgetErrorType) => {
      if (error.level === 'error') {
        setupFormError(error);
      }
      if (error.notificationText) {
        notify(error);
      }
    }, [notify, setupFormError]);

    const safeRequest = (opts: {}) => (
      request(opts)
        .then((response: Response) => {
          if (response.status === 504) {
            addError(CamundaError);
          }
          return response;
        }));

    return (
      <ErrorHandlingContext.Provider value={{
        addError,
        formError,
        safeRequest,
      }}>
        <WrappedComponent {...props} />
      </ErrorHandlingContext.Provider>
    );
  };

  return ErrorHandlingProvider;
};

export default ErrorHandlingContext;
