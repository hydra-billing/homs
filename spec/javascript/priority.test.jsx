import React from 'react';
import renderer from 'react-test-renderer';
import Priority from 'hbw/components/shared/element/priority';
import { withTranslationContext } from 'hbw/components/shared/context/translation';
import { withConnectionContext } from 'hbw/components/shared/context/connection';
import { waitFor } from '@testing-library/react';
import compose from 'shared/utils/compose';
import ConnectionMock from './helpers/mock_connection';

describe('Priority label should be rendered properly', () => {
  const PriorityWithContext = compose(
    withConnectionContext({ connection: ConnectionMock }),
    withTranslationContext({ locale: { code: 'en' } })
  )(Priority);

  const priorityComponent = priority => (
    renderer
      .create(<PriorityWithContext priority={priority} />)
      .toJSON()
  );

  const matchSnapshot = priorityName => priorityValue => (
    it(`for priority ${priorityValue} renders '${priorityName}'`, async () => {
      await waitFor(() => {
        expect(priorityComponent(priorityValue)).toMatchSnapshot();
      });
    })
  );

  matchSnapshot('Medium')(-50);
  matchSnapshot('Medium')(0);
  matchSnapshot('Medium')(49.9);
  matchSnapshot('High')(50);
  matchSnapshot('High')(62);
  matchSnapshot('High')(74.9);
  matchSnapshot('Urgent')(75);
  matchSnapshot('Urgent')(123456);
});
