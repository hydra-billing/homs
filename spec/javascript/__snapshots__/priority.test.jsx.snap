// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Priority label should be rendered properly for priority -50 renders 'Medium' 1`] = `
<span
  className="badge priority medium"
>
  Medium
</span>
`;

exports[`Priority label should be rendered properly for priority 0 renders 'Medium' 1`] = `
<span
  className="badge priority medium"
>
  Medium
</span>
`;

exports[`Priority label should be rendered properly for priority 49.9 renders 'Medium' 1`] = `
<span
  className="badge priority medium"
>
  Medium
</span>
`;

exports[`Priority label should be rendered properly for priority 50 renders 'High' 1`] = `
<span
  className="badge priority high"
>
  High
</span>
`;

exports[`Priority label should be rendered properly for priority 62 renders 'High' 1`] = `
<span
  className="badge priority high"
>
  High
</span>
`;

exports[`Priority label should be rendered properly for priority 74.9 renders 'High' 1`] = `
<span
  className="badge priority high"
>
  High
</span>
`;

exports[`Priority label should be rendered properly for priority 75 renders 'Urgent' 1`] = `
<span
  className="badge priority urgent"
>
  Urgent
</span>
`;

exports[`Priority label should be rendered properly for priority 123456 renders 'Urgent' 1`] = `
<span
  className="badge priority urgent"
>
  Urgent
</span>
`;
